pkg/jobmanager/status.go

```diff
@@ -38,131 +46,183 @@ var TargetRoutingEvents = []event.Name{
-			if *candidateTargetStatus.Target == *testEvent.Data.Target {
-				currentTargetStatus = &currentTestStepStatus.TargetStatus[index]
+		// Update the TargetStatus object associated to the Target.
+		// If there is no TargetStatus associated yet, append it
+		var targetStatus *job.TargetStatus
+		for index, candidateStatus := range targetStatuses {
+			if *candidateStatus.Target == *testEvent.Data.Target {
+				targetStatus = &targetStatuses[index]
 				break
 			}
 		}
-		if currentTargetStatus == nil {
+
+		if targetStatus == nil {
 			// There is no TargetStatus associated with this Target, create one
-			currentTestStepStatus.TargetStatus = append(
-				currentTestStepStatus.TargetStatus,
-				job.TargetStatus{Target: testEvent.Data.Target},
-			)
-			lastTargetStatus := len(currentTestStepStatus.TargetStatus) - 1
-			currentTargetStatus = &currentTestStepStatus.TargetStatus[lastTargetStatus]
+			targetStatuses = append(targetStatuses, job.TargetStatus{Target: testEvent.Data.Target})
+			targetStatus = &targetStatuses[len(targetStatuses)-1]
 		}
 
-		switch eventName := testEvent.Data.EventName; eventName {
-		case target.EventTargetIn:
-			currentTargetStatus.InTime = testEvent.EmitTime
-		case target.EventTargetOut:
-			currentTargetStatus.OutTime = testEvent.EmitTime
-		case target.EventTargetErr:
-			currentTargetStatus.OutTime = testEvent.EmitTime
-			currentTargetStatus.Error = testEvent.Data.Payload
+		if testEvent.Data.EventName == target.EventTargetIn {
+			targetStatus.InTime = testEvent.EmitTime
+		} else if testEvent.Data.EventName == target.EventTargetOut {
+			targetStatus.OutTime = testEvent.EmitTime
+		} else if testEvent.Data.EventName == target.EventTargetErr {
+			targetStatus.OutTime = testEvent.EmitTime
+			targetStatus.Error = testEvent.Data.Payload
 		}
 	}
 
-	return nil
+	return targetStatuses, nil
 }
 
-// buildTestStepStatus populates a TestStatus object with TestStep status information
-func (jm *JobManager) buildTestStepStatus(jobID types.JobID, currentTest *test.Test, currentTestStatus *job.TestStatus) error {
+// buildTestStepStatus builds the status object of a test step belonging to a test
+func (jm *JobManager) buildTestStepStatus(coordinates TestStepCoordinates) (*job.TestStepStatus, error) {
+
+	testStepStatus := job.TestStepStatus{TestStepName: coordinates.TestStepName, TestStepLabel: coordinates.TestStepLabel}
 
-	// Build a map of events that should not be rendered as part of the status
+	// Fetch all Events associated to this TestStep
+	// TODO: This query will have to include the runID, which can be found in the coordinates
+	testEvents, err := jm.testEvManager.Fetch(
+		[]testevent.QueryField{
+			testevent.QueryJobID(coordinates.JobID),
+			testevent.QueryTestName(coordinates.TestName),
+			testevent.QueryTestStepLabel(coordinates.TestStepLabel),
+		},
+	)
+	if err != nil {
+		return nil, fmt.Errorf("could not fetch events associated to test step %s: %v", coordinates.TestStepLabel, err)
```

Comment:
It would be better to use `%w` instead of `%v`.

Reply:
I agree - this could be incorporated in #25